### PR TITLE
Function tests cleanup (part-2)

### PR DESCRIPTION
**Pull request by kichalla**
This has the following changes:
- Making UrlHelperWebSite to a sample
- Deleting ResponseCacheWebSite and converting functional tests to unit tests
- Moving TryValidateModel and buddy type functional tests to integration tests.

I will send a separate PR to handle RemoteAttribute test changes.

@rynowak @dougbu

### CODE REVIEW

File: Mvc.NoFun.sln [hunk 1]
```diff
@@ -85,6 +85,8 @@ Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.AspNet.Mvc.TestDi
 EndProject
 Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "MvcSandbox", "samples\MvcSandbox\MvcSandbox.xproj", "{14ED4476-9F24-4776-8417-EA6927F6C9C9}"
 EndProject
+Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "UrlHelperSample", "samples\UrlHelperSample\UrlHelperSample.xproj", "{712A98A0-227F-40D9-9342-8FD178DACB26}"
```
Comment (dougbu) [inline Mvc.NoFun.sln]: Suggest we stick w/ the `Sample.Web` suffix i.e. `UrlHelperSample.Web`.

File: samples/UrlHelperSample/project.json [hunk 2]
```diff
@@ -1,12 +1,11 @@
 {
     "commands": {
-        "web": "Microsoft.AspNet.Hosting server=Microsoft.AspNet.Server.WebListener server.urls=http://localhost:5001",
-        "kestrel": "Microsoft.AspNet.Hosting --server Microsoft.AspNet.Server.Kestrel --server.urls http://localhost:5000"
+        "weblistener": "Microsoft.AspNet.Hosting server=Microsoft.AspNet.Server.WebListener server.urls=http://localhost:5001",
```
Comment (dougbu) [inline samples/UrlHelperSample/project.json]: Don't see a reason to stick w/ port 5001. Why isn't this just `Microsoft.AspNet.Server.WebListener`?

File: samples/UrlHelperSample/wwwroot/web.config [hunk 3]
```diff
@@ -0,0 +1,9 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (dougbu) [inline samples/UrlHelperSample/wwwroot/web.config]: Don't check this file in. It's needed only while debugging and VS recreates it.

File: samples/UrlHelperSample/project.json [hunk 4]
```diff
@@ -1,12 +1,11 @@
...
+        "web": "Microsoft.AspNet.Server.Kestrel"
     },
     "dependencies": {
-        "Microsoft.AspNet.Server.Kestrel": "1.0.0-*",
         "Microsoft.AspNet.Mvc": "6.0.0-*",
-        "Microsoft.AspNet.Mvc.TestConfiguration": "1.0.0",
+        "Microsoft.AspNet.Server.Kestrel": "1.0.0-*",
         "Microsoft.AspNet.Server.WebListener": "1.0.0-*",
         "Microsoft.AspNet.StaticFiles": "1.0.0-*"
```
Comment (dougbu) [inline samples/UrlHelperSample/project.json]: Why is this dependency required?

File: samples/UrlHelperSample/wwwroot/web.config (hunk 3 above)
Comment (dougbu) [inline samples/UrlHelperSample/wwwroot/web.config]: Further, if the whole directory can be left out, great.

File: samples/UrlHelperSample/project.json [hunk 5]
```diff
@@ -1,12 +1,11 @@
...
     },
```
Comment (dougbu) [inline samples/UrlHelperSample/project.json]: Remove the old `webroot` setting in this file. Should either be set in a `hosting.json` file (perhaps to `Content`) or shouldn't be set at all.

File: samples/UrlHelperSample/AppOptions.cs [hunk 6]
```diff
@@ -1,7 +1,7 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-namespace UrlHelperWebSite
+namespace UrlHelperSample
```
Comment (dougbu) [inline samples/UrlHelperSample/AppOptions.cs]: `namespace UrlHelperSample.Web`

File: test/WebSites/ActionConstraintsWebSite/ActionConstraintsWebSite.xproj [hunk 7]
```diff
@@ -1,18 +1,19 @@
...
+<Project ToolsVersion="14.0.23107" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
-    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">14.0</VisualStudioVersion>
+    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">14.0.23107</VisualStudioVersion>
     <VSToolsPath Condition="'$(VSToolsPath)' == ''">$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)</VSToolsPath>
   </PropertyGroup>
   <Import Project="$(VSToolsPath)\DNX\Microsoft.DNX.Props" Condition="'$(VSToolsPath)' != ''" />
   <PropertyGroup Label="Globals">
-    <ProjectGuid>af210f69-9d31-43af-ac3a-cd366e252218</ProjectGuid>
+    <ProjectGuid>e51c57e3-4d52-47a9-aba6-4e6fff027a5b</ProjectGuid>
```
Comment (dougbu) [inline test/WebSites/ActionConstraintsWebSite/ActionConstraintsWebSite.xproj]: Why is this file / project changing?

File: samples/UrlHelperSample/CustomUrlHelper.cs [hunk 8]
```diff
@@ -0,0 +1,66 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.Mvc;
+using Microsoft.AspNet.Mvc.Routing;
+
+namespace UrlHelperSample
```
Comment (dougbu) [inline samples/UrlHelperSample/CustomUrlHelper.cs]: Is anything other than the `namespace` different? Odd it's not showing up as a rename.

File: samples/UrlHelperSample/CustomUrlHelper.cs [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        {
+            _options = options;
+        }
+
+        /// <summary>
+        /// Depending on config data, generates an absolute url pointing to a CDN server
+        /// or falls back to the default behavior
+        /// </summary>
+        /// <param name="contentPath"></param>
+        /// <returns></returns>
```
Comment (dougbu) [inline samples/UrlHelperSample/CustomUrlHelper.cs]: Existing issue but should finish the doc comments here.

**Comment by dougbu** [comment]
Links shown in `UrlHelperSample` don't work anyway. So don't check in the `UrlHelperSample/Content/bootstrap.min.css` file or use `StaticFiles`.

File: test/Microsoft.AspNet.Mvc.Core.Test/ResponseCacheAttributeTest.cs [hunk 10]
```diff
@@ -181,6 +186,127 @@ public void CreateInstance_DoesNotThrowWhenTheDurationIsNotSet_WithNoStoreFalse(
...
+            StringValues values;
+            Assert.True(response.Headers.TryGetValue("Cache-Control", out values));
+            var data = Assert.Single(values);
+            AssertHeaderEquals("public, max-age=100", data);
+            Assert.True(response.Headers.TryGetValue("Vary", out values));
+            data = Assert.Single(values);
+            Assert.Equal("Accept", data);
+        }
+
+        public static IEnumerable<object[]> CacheControlData
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ResponseCacheAttributeTest.cs]: Use `TheoryData<ResponseCacheAttribute, string>`.

File: test/Microsoft.AspNet.Mvc.Core.Test/ResponseCacheAttributeTest.cs [hunk 11]
```diff
@@ -199,5 +325,22 @@ private IServiceProvider GetServiceProvider(Dictionary<string, CacheProfile> cac
 
             return serviceProvider.Object;
         }
+
+        private ActionExecutingContext GetActionExecutingContext(List<IFilterMetadata> filters = null)
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ResponseCacheAttributeTest.cs]: Why does the `filters` parameter have a default value? Always called with a value.

Any why `List<IFilterMetadata>` instead of the easier-to-use `params IFilterMetadata[]`?

File: test/Microsoft.AspNet.Mvc.Core.Test/ResponseCacheAttributeTest.cs [hunk 12]
```diff
@@ -199,5 +325,22 @@ private IServiceProvider GetServiceProvider(Dictionary<string, CacheProfile> cac
...
+            return new ActionExecutingContext(
+                new ActionContext(new DefaultHttpContext(), new RouteData(), new ActionDescriptor()),
+                filters ?? new List<IFilterMetadata>(),
+                new Dictionary<string, object>(),
+                new object());
+        }
+
+        private void AssertHeaderEquals(string expected, string actual)
+        {
+            // OrderBy is used because the order of the results may very depending on the platform / client.
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ResponseCacheAttributeTest.cs]: vary

File: test/Microsoft.AspNet.Mvc.Core.Test/ResponseCacheAttributeTest.cs [hunk 13]
```diff
@@ -181,6 +186,127 @@ public void CreateInstance_DoesNotThrowWhenTheDurationIsNotSet_WithNoStoreFalse(
...
+            };
+            var filter = (ResponseCacheFilter)responseCacheAttribute.CreateInstance(
+                GetServiceProvider(cacheProfiles: null));
+            var context = GetActionExecutingContext(new List<IFilterMetadata> { filter });
+
+            // Act & Assert
+            var exception = Assert.Throws<InvalidOperationException>(() =>
+                           {
+                               filter.OnActionExecuting(context);
+                           });
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ResponseCacheAttributeTest.cs]: Unrandomize the indentation.

File: samples/UrlHelperSample/wwwroot/web.config (hunk 3 above)
Comment (kichalla) [inline samples/UrlHelperSample/wwwroot/web.config]: But since its a sample we tend to use it in VS and if we do not check this in we would always see a pending change to be committed.

File: samples/UrlHelperSample/project.json (hunk 4 above)
Comment (kichalla) [inline samples/UrlHelperSample/project.json]: Sure, can be removed

File: samples/UrlHelperSample/CustomUrlHelper.cs (hunk 8 above)
Comment (kichalla) [inline samples/UrlHelperSample/CustomUrlHelper.cs]: There are no other changes here

File: samples/UrlHelperSample/wwwroot/web.config (hunk 3 above)
Comment (dougbu) [inline samples/UrlHelperSample/wwwroot/web.config]: These samples are rarely used from VS and my aim here is to get rid of the useless wwwroot directory entirely.

File: test/WebSites/ActionConstraintsWebSite/ActionConstraintsWebSite.xproj (hunk 7 above)
Comment (kichalla) [inline test/WebSites/ActionConstraintsWebSite/ActionConstraintsWebSite.xproj]: Looks like one of my commits deleted this project, but I added it back at which stage this got changed...I do not see any side effects of this change though...

File: test/Microsoft.AspNet.Mvc.Core.Test/ResponseCacheAttributeTest.cs (hunk 11 above)
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.Core.Test/ResponseCacheAttributeTest.cs]: :+1:

**Comment by kichalla** [comment]
:up: :date:

File: test/WebSites/ActionConstraintsWebSite/ActionConstraintsWebSite.xproj (hunk 7 above)
Comment (dougbu) [inline test/WebSites/ActionConstraintsWebSite/ActionConstraintsWebSite.xproj]: Churn for churn's sake will mess everyone else up. Just revert the change here and let VS update the `.sln` file again. Or revert the unnecessary changes in the `.sln` file.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/TryUpdateModelIntegrationTest.cs [hunk 14]
```diff
@@ -1112,5 +1117,7 @@ System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
                ModelBindingTestHelper.GetObjectValidator(),
                operationContext.ValidatorProvider);
         }
+
+
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/TryUpdateModelIntegrationTest.cs]: Looks like this file doesn't need to change at all.

**Comment by kichalla** [comment]
:up: :date: @dougbu

**Comment by kichalla** [comment]
ping

File: samples/UrlHelperSample.Web/project.json [hunk 15]
```diff
@@ -0,0 +1,15 @@
+{
+    "commands": {
+        "weblistener": "Microsoft.AspNet.Hosting server=Microsoft.AspNet.Server.WebListener server.urls=http://localhost:5001",
```
Comment (dougbu) [inline samples/UrlHelperSample.Web/project.json]: Change value to just be `"Microsoft.AspNet.Server.WebListener"`.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/TryValidateModelIntegrationTest.cs [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Mvc.Abstractions;
+using Microsoft.AspNet.Mvc.ModelBinding;
+using Microsoft.AspNet.Testing;
+using Microsoft.AspNet.Http.Internal;
+#if !DNXCORE50
+using Microsoft.AspNet.Testing.xunit;
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/TryValidateModelIntegrationTest.cs]: What is this for? Are all of these `using`s necessary?

File: test/Microsoft.AspNet.Mvc.IntegrationTests/CompanyNameAttribute.cs [hunk 17]
```diff
@@ -0,0 +1,18 @@
+﻿using System.ComponentModel.DataAnnotations;
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/CompanyNameAttribute.cs]: header

File: test/Microsoft.AspNet.Mvc.IntegrationTests/Models/SoftwareViewModel.cs [hunk 18]
```diff
@@ -0,0 +1,29 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.DataAnnotations;
+
+namespace Microsoft.AspNet.Mvc.IntegrationTests
+{
+    [ModelMetadataType(typeof(Software))]
+    public class SoftwareViewModel
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/Models/SoftwareViewModel.cs]: Adding a lot of very similar model types with repeated use of the same attributes. Scale things back to highlight the important parts of the tests.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ProductValidatorAttribute.cs [hunk 19]
```diff
@@ -0,0 +1,38 @@
+﻿using System.ComponentModel.DataAnnotations;
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ProductValidatorAttribute.cs]: header

File: test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs [hunk 20]
```diff
@@ -1,137 +0,0 @@
...
-using System.Net.Http;
-using System.Text;
-using System.Threading.Tasks;
-using Microsoft.AspNet.Testing;
-using Newtonsoft.Json;
-using Xunit;
-
-namespace Microsoft.AspNet.Mvc.FunctionalTests
-{
-    public class TryValidateModelTest : IClassFixture<MvcTestFixture<ValidationWebSite.Startup>>
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs]: Remove corresponding code from the `ValidationWebSite`. Just dangling bloat now.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/CustomUrlHelperTests.cs [hunk 21]
```diff
@@ -1,67 +0,0 @@
-// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/CustomUrlHelperTests.cs]: Huh? Where are the tests of `UrlHelperSample.Web`? Everything of interest in that site should still be tested.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/Models/SoftwareViewModel.cs (hunk 18 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/Models/SoftwareViewModel.cs]: And, if possible, make some of these classes `private` in whatever (single) test class they're used.

**Comment by dougbu** [comment]
:watch:

File: test/Microsoft.AspNet.Mvc.FunctionalTests/CustomUrlHelperTests.cs (hunk 21 above)
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/CustomUrlHelperTests.cs]: Good point. Will add these back

File: test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs (hunk 20 above)
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/TryValidateModelTest.cs]: Removed it as part of the part3 PR as RemoteAttribute used it in this PR

File: test/Microsoft.AspNet.Mvc.IntegrationTests/Models/SoftwareViewModel.cs (hunk 18 above)
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/Models/SoftwareViewModel.cs]: Currently these models are being used in two tests(TryValiateModel and ModelMetadataType) and soI did not create them as private classes.

**Comment by kichalla** [comment]
:up: :date:

**Comment by dougbu** [comment]
Late realization: PR is missing removal of at least `CustomUrlHelper`, `CustomUrlHelperFactory` and Startup` from the old `UrlHelperWebSite`. The new `UrlHelperSample.Web` project also needs a `readme.md` file.

**Comment by dougbu** [comment]
:watch: just to see diffs to the remaining files that actually have been renamed, not added.

**Comment by dougbu** [comment]
:shipit: once the commits here can be squashed together w/ those in #3699.

(Manually compared the missing files and found no unexpected changes.)

**Comment by kichalla** [comment]
57b88baad0372d3fe3f4bc8b9534d2643ddbf2dd
